### PR TITLE
kitty: add module

### PR DESCRIPTION
**Pull request by iv-nn**

### CODE REVIEW

File: modules/programs/kitty.nix [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    home.packages = [ pkgs.kitty ] ++ optionalPackage cfg.font;
+
+    xdg.configFile."kitty/kitty.conf".text = ''
+      # Generated by Home Manager.
+      # See https://sw.kovidgoyal.net/kitty/conf.html
+
+      font_family ${cfg.font.name}
```
Comment (teto) [inline modules/programs/kitty.nix]: font being null by default can cause problems

Comment (teto) [reply modules/programs/kitty.nix]: there is a fontType submodule in gtk.nix that oculd be reused.

File: modules/lib/font-type.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{ lib }:
+
+with lib;
+
+{
+  fontType = types.submodule {
```
Comment (rycee) [inline modules/lib/font-type.nix]: Can move this to `types.nix` instead.

File: modules/programs/kitty.nix [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  toKittyKeybindings = generators.toKeyValue {
+    mkKeyValue = key: command: "map ${key} ${command}";
+  };
+in
+
+{
+  options.programs.kitty = {
+    enable = mkEnableOption "Kitty terminal emulator";
+
+    config = mkOption {
```
Comment (rycee) [inline modules/programs/kitty.nix]: This option should be called `settings` instead for consistency with other modules and avoiding the already very overloaded `config` name.

File: modules/programs/kitty.nix [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        {
+          "ctrl+c" = "copy_or_interrupt";
+          "ctrl+f>2" = "set_font_size 20";
+        }
+      '';
+    };
+
+    extraConfig = mkOption {
+      default = "";
+      type = types.str;
```
Comment (rycee) [inline modules/programs/kitty.nix]: Should be of type `types.lines` to allow option merging.

File: modules/programs/kitty.nix [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      description = "Additional configuration to add.";
+      example = "";
```
Comment (rycee) [inline modules/programs/kitty.nix]: An empty example is not very useful, should be either removed or given some content.

File: modules/programs/kitty.nix [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  eitherStrBoolInt = with types; either str (either bool int);
+
+  optionalPackage = opt:
+    optional (opt != null && opt.package != null) opt.package;
+
+  fontType = (import ../lib/font-type.nix { inherit lib; }).fontType;
+
+  toKittyConfig = generators.toKeyValue {
+    mkKeyValue = key: value:
+    let
```
Comment (rycee) [inline modules/programs/kitty.nix]: Indent this block two spaces.

File: modules/programs/kitty.nix [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      value' =
+        if isBool value then (if value then "yes" else "no")
```
Comment (rycee) [inline modules/programs/kitty.nix]: Move `then` branch to next line

**Comment by rycee** [comment]
Thanks for the contribution, it looks generally quite good, just added a few comments.

File: modules/lib/font-type.nix (hunk 2 above)
Comment (iv-nn) [reply modules/lib/font-type.nix]: Should I be able to call it directly through `lib.hm.types` once moved to `types.nix`? I've only managed to call it like done [here](https://github.com/rycee/home-manager/blob/master/modules/programs/emacs.nix#L39).

Comment (rycee) [reply modules/lib/font-type.nix]: Yes, `lib.hm.types.fontType` should work. Like here: https://github.com/rycee/home-manager/blob/95c8007b8fe4433cd217db83ce26b8e29875ba06/modules/home-environment.nix#L235

Comment (iv-nn) [reply modules/lib/font-type.nix]: It works, I think my fork wasn't up to date when I last tried.

**Comment by iv-nn** [comment]
Should I change anything else?

**Comment by rycee** [comment]
Thanks! I squashed and extracted the font type changes and the kitty module into two separate commits. Also fixed the formatting that the CI job was erroring out on. Rebased to master in 9ab0d2305c059937f907a0d3689ac65891b447f3.

**Pull request closed without merging**